File: src/scene/encounter/Encounter.java
package scene.encounter;

import model.Party;

/**
 * An abstract encounter.  Encounters have a certain chance of occurring,
 * and are meant to execute their own action, as well as return a message
 * or scene encapsulated in a EncounterNotification
 */
public abstract class Encounter {
	protected int value;
	protected Party party;
	protected EncounterID id;
	
	public Encounter(Party party, int value, EncounterID id) {
		this.party = party;
		this.value = value;
		this.id = id;
	}
	
	public boolean isInRange(int num, int timeIndex) {
		return num < value * id.getFrequencies()[timeIndex];
	}
	
	public int getValue(int timeIndex) {
		return value * id.getFrequencies()[timeIndex];
	}
	
	/**
	 * Executes the logic of the encounter.
	 * @return An EncounterNotification describing the encounter.
	 */
	public abstract EncounterNotification doEncounter();
	
	/**
	 * Creates the notification data for the Encounter.
	 * @return An EncounterNotification with the
	 */
	protected abstract EncounterNotification makeNotification();
	
	public abstract EncounterID getEncounterID();

	public void increaseValue(int amount) {
		value += amount;
	}
}

File: web/src/scene/encounter/Encounter.ts
import { Party } from "../../model/Party";
import { EncounterNotification } from "./EncounterNotification";
import { EncounterID, getEncounterFrequencies } from "./EncounterID";

export abstract class Encounter {
  protected value: number;
  protected party: Party;
  protected id: EncounterID;

  constructor(party: Party, value: number, id: EncounterID) {
    this.party = party;
    this.value = value;
    this.id = id;
  }

  isInRange(num: number, timeIndex: number): boolean {
    return num < this.value * getEncounterFrequencies(this.id)[timeIndex];
  }

  getValue(timeIndex: number): number {
    return this.value * getEncounterFrequencies(this.id)[timeIndex];
  }

  abstract doEncounter(): EncounterNotification;

  protected abstract makeNotification(): EncounterNotification;

  getEncounterID(): EncounterID {
    return this.id;
  }

  increaseValue(amount: number): void {
    this.value += amount;
  }
}
